### ISSUE TITLE
Question: How do you compile a flat list of instructions?

### ISSUE THREAD

**Issue by Heath123** [WebAssembly/binaryen#4737]
The page https://github.com/WebAssembly/binaryen/wiki/Compiling-to-WebAssembly-with-Binaryen says:
> If, instead, you have a more "flat" representation of simple instructions in lists in basic blocks, then you can pass that in as well, just create Blocks and fill them with such simple instructions. That will use more local variables, but the Binaryen optimizer can optimize those out.

However, e.g. `BinaryenBinary` takes `left` and `right` parameters, and passing in `null` results in:
```
void wasm::Binary::finalize(): Assertion `left && right' failed.
```
So how would I make a module from a list of WebAssembly instructions?

**Comment by Heath123** [comment]
Should this be a discussion?

**Comment by kripken** [comment]
Good point, that text isn't very clear. I edited it now, and added an example - does that help?

(This could be a discussion, yeah - we aren't very strict about that here though.)

**Comment by Heath123** [comment]
Ah, thanks, I see what that meant now - still using the tree structure but storing and loading locals each time. That wasn't quite what I was looking for though, I wanted to generate the normal stack machine instructions so instead of
```wat
 (i32.add
  (local.get $0)
  (local.get $1)
 )
```
I could do the normal
```wat
(local.get $0)
(local.get $1)
(i32.add)
```
Is that possible as well? I know BInaryen doesn't use that format internally but it has to be able to understand it to read WAT files, is there an API for it?

**Comment by kripken** [comment]
There isn't an API for that atm. Binaryen can't even read wat in that format, actually. At least the wat side may get fixed with @tlively 's current work on a new parser, though.

Are you using text, or the C API, or something else?

**Comment by Heath123** [comment]
I’d probably have to convert it out of the stack machine format anyway though because Python bytecode is a lot less strict so maybe that’s not a problem 

I have no idea if this will actually be possible but it’s a fun hobby project if I can get it working for simple code I guess

**Comment by kripken** [comment]
Sounds like a cool project!

I hope adding locals and local operations wouldn't be too hard to do. But eventually it might be nice to add a C API for real stack machine input, where Binaryen would automatically convert it to use locals and local operations.

**Comment by tlively** [comment]
We now have the `IRBuilder` C++ API that is used by both the text and binary parsers to translate streams of stack machine instructions to structured Binaryen IR, but it's not (yet) exposed in the C API. If that would be useful, please open a new feature request issue. I'll close this issue because the initial question has been answered.
